tsdb/store.go

```diff
@@ -373,41 +373,64 @@ func (s *Store) ShardIteratorCreator(id uint64) influxql.IteratorCreator {
 	return &shardIteratorCreator{sh: sh}
 }
 
+// shardExecuteFn checks if a function should be run on a shard and returns
+// a function to execute on that shard if it should.
+type shardExecuteFn func(shardID uint64, sh *Shard) func() error
```

Comment:
It's a bit confusing, but the reason is because the initial filter needs cannot be parallelized as the executor needs to modify the wait group if the condition is true. The function returned is the function run in a goroutine.